### ISSUE TITLE
Incorrect __init__ arguments when overriding @dataclass field to a ClassVar

### ISSUE THREAD

**Issue by petrusek** [microsoft/pyright#1854]
**Describe the bug**
I've noticed pyright and python inconsistency - what works in one, doesn't in the other. I hope my code is OK (I'm relatively new to python3).

**To Reproduce**
It is a special case of overriding dataclass attributes from e.g. `int` to `ClassVar[int]` in a subclass... which removes them from the autogenerated `__init__` function.

**Expected behavior**
Properly removing ClassVars from the subclasses `__init__` argument list.

**Screenshots or Code**
```python
from dataclasses import dataclass
from typing import ClassVar

@dataclass
class Base:
    x: int
    y: int

@dataclass
class Special(Base):
    x: ClassVar[int] = 1
    z: int

# all OK
print(Base(x=1, y=2))  # Prints: Base(x=1, y=2)

# Pylance error (Argument missing for parameter "x")
# Python OK
s = Special(y=2, z=3)
print(s.x, s)  # Prints: 1 Special(y=2, z=3)

# Pylance error (Argument missing for parameter "z")
# Python OK
s = Special(2, 3)
print(s.x, s)  # Prints: 1 Special(y=2, z=3)

# Pylance OK
# Pyton error (unexpected keyword argument 'x')
print(Special(x=1, y=2, z=3))

# Pylance OK
# Python error (takes 3 positional arguments but 4 were given)
print(Special(1, 2, 3))

```

**VS Code extension or command-line**
I'm running pyright as a VS Code extension via Pylance version v2021.5.1
Python version Python 3.9.4

**Additional context**
Might be related to #1804 ... I've tried turning on `reportIncompatibleVariableOverride` but I didn't see any additional errors (I guess `int` and `ClassVar[int]` should be compatible).

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.139, which I've just published. It will also be included in the next release of pylance.

**Comment by petrusek** [comment]
Hello @erictraut,
sorry for posting to a closed issue, but I've noticed a situation today, when the fixed version is still giving me an error. It happens when there are multiple subclasses (in a chain) and each of them redeclares a different class variable.

```python
from dataclasses import dataclass
from typing import ClassVar

@dataclass
class Base:
    x: int
    y: int

@dataclass
class Special(Base):
    x: ClassVar[int] = 1
    z: int

@dataclass
class VerySpecial(Special):
    y: ClassVar[int] = 2


# all OK
print(Base(x=1, y=2))  # Prints: Base(x=1, y=2)

# all OK now
s = Special(y=2, z=3)
print(s.x, s)  # Prints: 1 Special(y=2, z=3)

# Pylance error (Argument missing for parameter "x")
# Python OK
s = VerySpecial(z=3)
print(s.x, s.y, s)  # Prints: 1 2 VerySpecial(z=3)

```


**VS Code extension or command-line**
I'm running pyright as a VS Code extension via Pylance version v2021.5.4
Python version Python 3.9.4

**Comment by erictraut** [comment]
Thanks for the bug report. I hadn't considered this case. I've updated the logic to accommodate this case.

What you're doing here is arguably a type consistency error. Overriding a class variable with an instance variable or vice versa should be considered an error condition. Mypy reports it as such. I've updated the "reportIncompatibleVariableOverride" diagnostic check to report this case, so this code will now generate errors in strict mode.

**Comment by petrusek** [comment]
> What you're doing here is arguably a type consistency error.

Oh, I didn't realize that... I though if it is all `int` and I'm only specifying it "closer", it should be fine, but you are right - instance vs class variables are something different.... thank you for pointing it out. But since we are at it, I would like to ask your opinion about how I use it.

My more specific usecase is this:
I'm parsing binary messages with a data field that identifies the message type. Some messages, let's say event messages have their own structure and their own data field for more specific identification... so I wanted to depict this hierarchy with data classes, so I could use the parent class everywhere where I receive more types of messages and inside the function I usually use `isinstance` or eventually I could compare the identification field (msg_type or event_id).

```python
@dataclass(frozen=True)
class Message:  # base message
    msg_type: int

@dataclass(frozen=True)
class SomeMessage(Message):  # all other subtypes have specific msg_type
    msg_type: ClassVar[int] = 1
    payload: float  # message specific payload, could be anything

@dataclass(frozen=True)
class EventMessage(Message):  # base event
    msg_type: ClassVar[int] = 2  # all msg_type == 2 are event messages
    event_id: int  # we know it has an event_id

@dataclass(frozen=True)
class SomeEventMessage(EventMessage):  # locked in msg_type == 1 and event_id == 1
    event_id: ClassVar[int] = 1
    payload_a: int  # message specific payload
    payload_b: float

@dataclass(frozen=True)
class OtherEventMessage(EventMessage):  # locked in msg_type == 1 and event_id == 2
    event_id: ClassVar[int] = 2
    payload_x: int  # message specific payload
    payload_y: float

```

I had a deeper look at dataclasses, whether they provide similar functionality without using `ClassVar`, and I've found
```python
msg_type: int = field(default=1, init=False)
```
which seems to work great.

I also wanted to use these values as constants, like `x == SomeMessage.msg_type`, which would fit the `ClassVar` meaning. It seems it works the same with `field(default=1)` - it allows to use it that way too (the field is defined on the class). In the end I use only `isinstance(msg, SomeMessage)` which I find to be the clearest (but is probably slower).


So my question is,
- should I use `field(default=1, init=False)`?
- is usage of `SomeMessage.msg_type` correct with only the `field(default=1)` without updating the type hint to ClassVar?
- any other ideas (am I going at least a bit in the right direction?)


Thank you very much for all your help... and also for pyright/pylance, they work great and without them I would be completely lost.

**Comment by erictraut** [comment]
You have a number of options depending on your objectives.

You can use a field definition with a default value. This means that each message instance will have a copy of the `msg_type` value rather than a shared value at the class level, so it will consume more memory and time to construct. Probably not a problem unless you instantiate millions of messages.

You can also use a ClassVar as long as you use it consistently in the class hierarchy. In the sample above, you're still overriding instance variables with class variables (`msg_type` is an instance variable in `Message` but is a class variable in the subclasses).

Another variant is to use a "tagged union", also sometimes called a "discriminating union". This is an approach I use extensively within the pyright source code (which is TypeScript, but the concept is the same). The idea here is to define a union of all message types and use a specific field (a class variable) to discriminate between these types. The field type needs to have an explicit literal type annotation. If you do this, the type checker can perform type narrowing based on comparisons to the discriminating field.

Here's how this approach would look:

```python
from dataclasses import dataclass
from typing import ClassVar, Literal
from enum import IntEnum

class MessageType(IntEnum):
    MESSAGE_TYPE_OTHER = 1
    MESSAGE_TYPE_EVENT = 2

class EventID(IntEnum):
    EVENT_ID_SOME = 1
    EVENT_ID_OTHER = 2

@dataclass(frozen=True)
class MessageBase:
    msg_type: ClassVar[MessageType]
    payload: float

@dataclass(frozen=True)
class SomeMessage(MessageBase):
    msg_type: ClassVar[
        Literal[MessageType.MESSAGE_TYPE_OTHER]
    ] = MessageType.MESSAGE_TYPE_OTHER
    payload: float

@dataclass(frozen=True)
class EventMessageBase(MessageBase):
    msg_type: ClassVar[
        Literal[MessageType.MESSAGE_TYPE_EVENT]
    ] = MessageType.MESSAGE_TYPE_EVENT
    event_id: ClassVar[EventID]

@dataclass(frozen=True)
class SomeEventMessage(EventMessageBase):
    event_id: ClassVar[Literal[EventID.EVENT_ID_SOME]] = EventID.EVENT_ID_SOME
    payload_a: int
    payload_b: float

@dataclass(frozen=True)
class OtherEventMessage(EventMessageBase):
    event_id: ClassVar[Literal[EventID.EVENT_ID_OTHER]] = EventID.EVENT_ID_OTHER
    payload_x: int
    payload_y: float

EventMessage = SomeEventMessage | OtherEventMessage
Message = SomeMessage | EventMessage

def func1(msg: Message):
    reveal_type(msg)  # SomeMessage | SomeEventMessage | OtherEventMessage

    if msg.msg_type == MessageType.MESSAGE_TYPE_EVENT:
        reveal_type(msg)  # SomeEventMessage | OtherEventMessage

        if msg.event_id == EventID.EVENT_ID_OTHER:
            reveal_type(msg)  # OtherEventMessage
            print(msg.payload_x)
        else:
            reveal_type(msg)  # SomeEventMessage
            print(msg.payload_a)
```

**Comment by petrusek** [comment]
Thank you very much for the in-depth explanation.
To tell the truth, my original goal was (something like) the discriminating union, but I was missing the "secret sauce" so it didn't work as I wanted... but now, after I saw your code it is a quite "obvious sauce" :-)... I was missing the correct union part
```python
EventMessage = SomeEventMessage | OtherEventMessage
Message = SomeMessage | EventMessage
```
... and I see now that the correct typehint is `ClassVar[Literal[...]]`.

Wonderful, suddenly it all makes sense. Thank you again.
